mdtoc.go

```diff
@@ -1,7 +1,192 @@
 package mdtoc
 
-import "io"
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"unicode"
+)
 
-func Generate(input io.Reader, output io.Writer) {
-	io.Copy(output, input)
+const headerFormat = "- [%s](#%s)"
+const atxHeader = "#"
+const headerIdent = "    "
+
+func isValidHeaderRune(r rune) bool {
+	return unicode.IsNumber(r) || unicode.IsLetter(r) || unicode.IsSpace(r)
+}
+
+func normalizeHeader(header string) string {
+	lowerNoHash := strings.TrimLeft(strings.ToLower(header), "#")
+	noInvalidChars := []rune{}
+
+	for _, r := range lowerNoHash {
+		if isValidHeaderRune(r) {
+			noInvalidChars = append(noInvalidChars, r)
+		}
+	}
+
+	return strings.Replace(string(noInvalidChars), " ", "-", -1)
+}
+
+type writer func(data string)
+
+func writeHeader(
+	writeOutput writer,
+	level int,
+	header string,
+	headersCount map[string]int,
+) {
+	normalizedHeader := normalizeHeader(header)
+	count := headersCount[normalizedHeader]
+	headersCount[normalizedHeader] = count + 1
+	if count > 0 {
+		normalizedHeader = fmt.Sprintf("%s-%d", normalizedHeader, count)
+	}
+	line := fmt.Sprintf(
+		headerFormat,
+		header,
+		normalizedHeader,
+	)
+	for i := 1; i < level; i++ {
+		writeOutput(headerIdent)
+	}
+	writeOutput(line + "\n")
+}
+
+func parseHeader(line string) (int, string, bool) {
+	if !startsWithAtxHeader(line) {
+		return 0, "", false
+	}
+	spaceTrimmed := strings.TrimRight(line, " ")
+	parsed := strings.Split(spaceTrimmed, " ")
+	if len(parsed) == 1 {
+		return 0, "", false
+	}
+	headerlevel := len(parsed[0])
+	header := parsed[1:]
+	return headerlevel, strings.Join(header, " "), true
+}
+
+func startsWithAtxHeader(line string) bool {
+	return strings.Index(line, atxHeader) == 0
+}
+
+func skipUntil(scanner *bufio.Scanner, stop func(string) bool) error {
+	for scanner.Scan() {
+		if stop(scanner.Text()) {
+			return nil
+		}
+	}
+	return errors.New("skipped all content")
+}
+
+func Generate(input io.Reader, output io.Writer) error {
+	headerStart := "<!-- mdtocstart -->"
+	tocHeader := "# Table of Contents"
+	headerEnd := "<!-- mdtocend -->"
+	scanner := bufio.NewScanner(input)
+	headersCount := map[string]int{}
+
+	var writeErr error
+	writeOutput := func(s string) {
+		if writeErr != nil {
+			return
+		}
+		_, writeErr = output.Write([]byte(s))
+	}
+
+	var original bytes.Buffer
+	var isCodeSection bool
+	var wroteHeader bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.TrimSpace(line) == headerStart {
+			err := skipUntil(scanner, func(l string) bool {
```

Comment:
Yeah, I have the case where I trim the spaces:

```
			err := skipUntil(scanner, func(l string) bool {
				return strings.TrimSpace(l) == headerEnd
			})
```

I could do a skipper that accepts just a string and always trim spaces, but if a document has the header and some lines with spaces only I would destroy those lines. Right now I'm just removing all newlines until I get to the text, but considering leaving spaces.